qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -445,61 +422,53 @@ def _wait_for_job(self, timeout=60, wait=5):
+        while self.status() not in JOB_FINAL_STATES:
             elapsed_time = time.time() - start_time
             if timeout is not None and elapsed_time >= timeout:
                 raise TimeoutError('QISKit timed out')
-            logger.info('status = %s (%d seconds)', api_result['status'],
-                        elapsed_time)
-
-            if 'status' not in api_result:
-                self._exception = QISKitError("get_job didn't return status: %s" %
-                                              (pprint.pformat(api_result)))
-                raise QISKitError("get_job didn't return status: %s" %
-                                  (pprint.pformat(api_result)))
 
+            logger.info('status = %s (%d seconds)', self._status, elapsed_time)
             time.sleep(wait)
-            api_result = self._update_status()
 
         if self.cancelled:
-            job_result = {'id': self._id, 'status': 'CANCELLED',
-                          'result': 'job cancelled'}
-            return Result(job_result)
-
-        elif self.exception:
-            job_result = {'id': self._id, 'status': 'ERROR',
-                          'result': str(self.exception)}
-            return Result(job_result)
-
-        if api_result is None:
-            api_result = self._api.get_job(self._id)
+            return Result({'id': self._id, 'status': 'CANCELLED',
+                           'result': 'job cancelled'})
 
+        job_data = self._api.get_job(self._id)
         job_result_list = []
-        for circuit_result in api_result['qasms']:
+        for circuit_result in job_data['qasms']:
             this_result = {'data': circuit_result['data'],
                            'name': circuit_result.get('name'),
                            'compiled_circuit_qasm': circuit_result.get('qasm'),
                            'status': circuit_result['status']}
             if 'metadata' in circuit_result:
                 this_result['metadata'] = circuit_result['metadata']
+
             job_result_list.append(this_result)
-        job_result = {'id': self._id,
-                      'status': api_result['status'],
-                      'used_credits': api_result.get('usedCredits'),
-                      'result': job_result_list}
-        job_result['backend_name'] = self.backend_name
-        return Result(job_result)
 
+        return Result({'id': self._id,
+                       'status': job_data['status'],
+                       'used_credits': job_data.get('usedCredits'),
+                       'result': job_result_list,
+                       'backend_name': self.backend_name})
+
+    def _wait_for_submitting(self):
+        """ Waits for the request to return a Job ID """
+        if self._id is None:
+            if self._future is None:
+                raise JobError("You have to submit before asking for status or results!")
+            submit_info = self._future.result(timeout=60)
```

Comment:
Would it be needed to intercept the potential [`TimeoutError`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result) that might be raised by this line if the timeout is reached, even if unlikely? It seems it can ripple through the upwards and for example end up raising an uncontroled Exception when checking the `.id` property?

Reply:
A very common use-case could be that the user would like to distinguish between a JobError() and a Timeout(), as the latter might be a temporary problem that can be solved by retrying. 
A solution could be re-raising a `JobTimeoutError` (which will inherit from JobError, so QISKitError), this will give the semantic meaning the user needs, and comply with our exception policy: "everything thrown from the SDK must inherit from QISKitError". 
Does it make sense?

Reply:
Sounds good - however there seems it still might be possible that a property call raises an exception (ie. a call to `job.id`)? I think we tend to avoid it if possible, so if it is a concern, it might be just a matter of intercepting it during `.id(self)`.

Reply:
Yeah, we will probably end up removing all properties and keep everything as methods, we are just discussing this right now.